File: lib/api/geo.dart
import 'dart:math';

/// 中国测绘局加密坐标。校内课堂考勤 / 公寓打卡的围栏都是 GCJ-02。
/// GPS 芯片和 AOSP fused 给出 WGS-84，不转换会在成都偏大约 300–500 米。
/// 高德点和国产 ROM 的 `network` 已经是 GCJ-02，不能再转。

const _a = 6378245.0;
const _ee = 0.00669342162296594323;
const _earth = 6371000.0;

bool geoOutOfChina(double lat, double lng) {
  return lng < 72.004 || lng > 137.8347 || lat < 0.8293 || lat > 55.8271;
}

/// 签到 / 打卡提交前：WGS-84 转 GCJ-02，已经是 GCJ 的点原样返回。
(double lat, double lng) campusGcj02(
  double lat,
  double lng, {
  required String source,
  String? datum,
  String? provider,
}) {
  if (!geoIsWgs84(source: source, datum: datum, provider: provider)) {
    return (lat, lng);
  }
  return wgs84ToGcj02(lat, lng);
}

/// [datum] 优先。否则 GPS / fused / 名字里带 fused 的系统点当 WGS-84。
bool geoIsWgs84({required String source, String? datum, String? provider}) {
  final d = datum?.trim().toLowerCase();
  if (d == 'gcj02') return false;
  if (d == 'wgs84') return true;
  final p = provider?.trim().toLowerCase() ?? '';
  if (p == 'network') return false;
  if (p == 'gps' ||
      p == 'passive' ||
      p.contains('fused') ||
      (p.contains('gps') && !p.contains('network'))) {
    return true;
  }
  switch (source) {
    case 'amap':
    case 'network':
    case 'demo':
      return false;
    case 'gps':
    case 'fused':
      return true;
    default:
      return false;
  }
}

/// 迭代反解。教室 GCJ 用来模拟芯片 GPS 时用。
(double lat, double lng) gcj02ToWgs84(double lat, double lng) {
  if (!lat.isFinite || !lng.isFinite) return (lat, lng);
  if (geoOutOfChina(lat, lng)) return (lat, lng);
  var wgsLat = lat;
  var wgsLng = lng;
  for (var i = 0; i < 8; i++) {
    final g = wgs84ToGcj02(wgsLat, wgsLng);
    wgsLat -= g.$1 - lat;
    wgsLng -= g.$2 - lng;
  }
  return (wgsLat, wgsLng);
}

(double lat, double lng) wgs84ToGcj02(double lat, double lng) {
  if (!lat.isFinite || !lng.isFinite) return (lat, lng);
  if (geoOutOfChina(lat, lng)) return (lat, lng);
  final dLat = _transformLat(lng - 105.0, lat - 35.0);
  final dLng = _transformLng(lng - 105.0, lat - 35.0);
  final rad = lat / 180.0 * pi;
  var magic = sin(rad);
  magic = 1 - _ee * magic * magic;
  final sqrtMagic = sqrt(magic);
  final nlat = (dLat * 180.0) / ((_a * (1 - _ee)) / (magic * sqrtMagic) * pi);
  final nlng = (dLng * 180.0) / (_a / sqrtMagic * cos(rad) * pi);
  return (lat + nlat, lng + nlng);
}

double geoMeters(double lat1, double lng1, double lat2, double lng2) {
  final p1 = lat1 * pi / 180;
  final p2 = lat2 * pi / 180;
  final dP = (lat2 - lat1) * pi / 180;
  final dL = (lng2 - lng1) * pi / 180;
  final a =
      sin(dP / 2) * sin(dP / 2) + cos(p1) * cos(p2) * sin(dL / 2) * sin(dL / 2);
  return 2 * _earth * atan2(sqrt(a), sqrt(1 - a));
}

double _transformLat(double x, double y) {
  var ret =
      -100.0 +
      2.0 * x +
      3.0 * y +
      0.2 * y * y +
      0.1 * x * y +
      0.2 * sqrt(x.abs());
  ret += (20.0 * sin(6.0 * x * pi) + 20.0 * sin(2.0 * x * pi)) * 2.0 / 3.0;
  ret += (20.0 * sin(y * pi) + 40.0 * sin(y / 3.0 * pi)) * 2.0 / 3.0;
  ret += (160.0 * sin(y / 12.0 * pi) + 320 * sin(y * pi / 30.0)) * 2.0 / 3.0;
  return ret;
}

double _transformLng(double x, double y) {
  var ret =
      300.0 + x + 2.0 * y + 0.1 * x * x + 0.1 * x * y + 0.1 * sqrt(x.abs());
  ret += (20.0 * sin(6.0 * x * pi) + 20.0 * sin(2.0 * x * pi)) * 2.0 / 3.0;
  ret += (20.0 * sin(x * pi) + 40.0 * sin(x / 3.0 * pi)) * 2.0 / 3.0;
  ret += (150.0 * sin(x / 12.0 * pi) + 300.0 * sin(x / 30.0 * pi)) * 2.0 / 3.0;
  return ret;
}

File: android/app/src/main/kotlin/cn/edu/swun/swun_ehall/data/geo/Geo.kt
package cn.edu.swun.swun_ehall.data.geo

import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

/**
 * Campus check-in fences are GCJ-02. GPS/fused chips emit WGS-84.
 * Amap and CN-ROM network are already GCJ-02.
 */
object Geo {
    private const val A = 6378245.0
    private const val EE = 0.00669342162296594323
    private const val EARTH = 6371000.0

    fun outOfChina(lat: Double, lng: Double): Boolean {
        return lng < 72.004 || lng > 137.8347 || lat < 0.8293 || lat > 55.8271
    }

    fun isWgs84(source: String, datum: String? = null, provider: String? = null): Boolean {
        val d = datum?.trim()?.lowercase()
        if (d == "gcj02") return false
        if (d == "wgs84") return true
        val p = provider?.trim()?.lowercase().orEmpty()
        if (p == "network") return false
        if (p == "gps" ||
            p == "passive" ||
            p.contains("fused") ||
            (p.contains("gps") && !p.contains("network"))
        ) {
            return true
        }
        return when (source) {
            "amap", "network", "demo" -> false
            "gps", "fused" -> true
            else -> false
        }
    }

    fun campusGcj02(
        lat: Double,
        lng: Double,
        source: String,
        datum: String? = null,
        provider: String? = null,
    ): Pair<Double, Double> {
        if (!isWgs84(source, datum, provider)) return lat to lng
        return wgs84ToGcj02(lat, lng)
    }

    fun gcj02ToWgs84(lat: Double, lng: Double): Pair<Double, Double> {
        if (!lat.isFinite() || !lng.isFinite()) return lat to lng
        if (outOfChina(lat, lng)) return lat to lng
        var wgsLat = lat
        var wgsLng = lng
        repeat(8) {
            val g = wgs84ToGcj02(wgsLat, wgsLng)
            wgsLat -= g.first - lat
            wgsLng -= g.second - lng
        }
        return wgsLat to wgsLng
    }

    fun wgs84ToGcj02(lat: Double, lng: Double): Pair<Double, Double> {
        if (!lat.isFinite() || !lng.isFinite()) return lat to lng
        if (outOfChina(lat, lng)) return lat to lng
        val dLat = transformLat(lng - 105.0, lat - 35.0)
        val dLng = transformLng(lng - 105.0, lat - 35.0)
        val rad = lat / 180.0 * PI
        var magic = sin(rad)
        magic = 1 - EE * magic * magic
        val sqrtMagic = sqrt(magic)
        val nlat = (dLat * 180.0) / ((A * (1 - EE)) / (magic * sqrtMagic) * PI)
        val nlng = (dLng * 180.0) / (A / sqrtMagic * cos(rad) * PI)
        return (lat + nlat) to (lng + nlng)
    }

    fun meters(lat1: Double, lng1: Double, lat2: Double, lng2: Double): Double {
        val p1 = lat1 * PI / 180
        val p2 = lat2 * PI / 180
        val dP = (lat2 - lat1) * PI / 180
        val dL = (lng2 - lng1) * PI / 180
        val a = sin(dP / 2) * sin(dP / 2) + cos(p1) * cos(p2) * sin(dL / 2) * sin(dL / 2)
        return 2 * EARTH * atan2(sqrt(a), sqrt(1 - a))
    }

    private fun transformLat(x: Double, y: Double): Double {
        var ret = -100.0 + 2.0 * x + 3.0 * y + 0.2 * y * y + 0.1 * x * y + 0.2 * sqrt(abs(x))
        ret += (20.0 * sin(6.0 * x * PI) + 20.0 * sin(2.0 * x * PI)) * 2.0 / 3.0
        ret += (20.0 * sin(y * PI) + 40.0 * sin(y / 3.0 * PI)) * 2.0 / 3.0
        ret += (160.0 * sin(y / 12.0 * PI) + 320 * sin(y * PI / 30.0)) * 2.0 / 3.0
        return ret
    }

    private fun transformLng(x: Double, y: Double): Double {
        var ret = 300.0 + x + 2.0 * y + 0.1 * x * x + 0.1 * x * y + 0.1 * sqrt(abs(x))
        ret += (20.0 * sin(6.0 * x * PI) + 20.0 * sin(2.0 * x * PI)) * 2.0 / 3.0
        ret += (20.0 * sin(x * PI) + 40.0 * sin(x / 3.0 * PI)) * 2.0 / 3.0
        ret += (150.0 * sin(x / 12.0 * PI) + 300.0 * sin(x / 30.0 * PI)) * 2.0 / 3.0
        return ret
    }
}
